### PR TITLE
Feature/environment propagate

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: (Feature | Fix | Bugfix): Describe here your pull request
Docs: https://github.com/conan-io/docs/pull/XXXX

- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [ ] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conan/tools/env/envtoolchain.py [hunk 1]
```diff
@@ -8,13 +8,23 @@ def __init__(self, conanfile):
...
-        return env
+        deps_env = Environment()
+        for dep in self._conanfile.dependencies.all:
+            dep_env = dep.new_env
+            if dep_env is not None:
+                deps_env = deps_env.compose(dep_env)
+
+        profile_env = Environment()
+        for k, v in self._conanfile._conan_env_values.data.items():
+            if k is None:
```
Comment (lasote) [inline conan/tools/env/envtoolchain.py]: what is these `k` being `None`? I don't remember what it means.

File: conan/tools/env/envtoolchain.py [hunk 2]
```diff
@@ -8,13 +8,23 @@ def __init__(self, conanfile):
...
+                for var, value in v.items():
+                    profile_env[var].prepend_path(value)
```
Comment (lasote) [inline conan/tools/env/envtoolchain.py]: What is doing the `prepend_path`?

File: conans/model/conan_file.py [hunk 3]
```diff
@@ -145,6 +146,7 @@ def __init__(self, output, runner, display_name="", user=None, channel=None):
         self._conan_using_build_profile = False
 
         self.layout = Layout()
+        self.dependencies = ConanFileDependencies()
```
Comment (lasote) [inline conans/model/conan_file.py]: Why this new model? Trying to optimize something?

File: conans/model/conanfile_interface.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+
+
+class ConanFileInterface:
```
Comment (lasote) [inline conans/model/conanfile_interface.py:3]: What for?

File: conans/model/conanfile_interface.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+    def __init__(self, conanfile):
+        self._conanfile = conanfile
+
+    @property
+    def new_env(self):
```
Comment (lasote) [inline conans/model/conanfile_interface.py]: new env?

File: conans/model/conanfile_interface.py (hunk 4 above)
Comment (memsharded) [reply conans/model/conanfile_interface.py:3]: To restrict the visibility that consumers have to their dependencies. Otherwise consumers might be able to call any method in the dependency conanfile or access any attribute.

File: conans/model/conanfile_interface.py (hunk 5 above)
Comment (memsharded) [reply conans/model/conanfile_interface.py]: Yes, it is a placeholder for a better name. The problem is ``environment()`` is the method name, and ``self.env`` already exists in conanfile, to contain the aggregated EnvInfo.

File: conans/model/conan_file.py (hunk 3 above)
Comment (memsharded) [reply conans/model/conan_file.py]: To give access to conanfiles to their dependencies. Navigating their dependencies one conanfile should be able:
- To gather information about cpp, that is, to compute the equivalent to ``deps_cpp_info`` from every dependency ``cpp_info``
- To gather information about dependencies references, including package_ids and recipe revisions.
- To gather information about real options of dependencies. Instead of ``self.options["mydep"]``, maybe it is ``self.dependencies["mydep"].options``. 

In summary, to replace the aggregated current view ``deps_cpp_info``, ``options``, etc, to a real full "graph" view, that can be aggregated on demand only.

File: conan/tools/env/envtoolchain.py (hunk 2 above)
Comment (memsharded) [reply conan/tools/env/envtoolchain.py]: PATH=MyValues;$PATH

As opposed to append, that does:

PATH=$PATH;MyValues

File: conan/tools/env/envtoolchain.py (hunk 1 above)
Comment (memsharded) [reply conan/tools/env/envtoolchain.py]: That is the current ``EnvValues`` model: {pattern: {var-name: var-value}}. The pattern is the package-name pattern, being ``None``= Affect all packages.

File: conans/model/conan_file.py (hunk 3 above)
Comment (lasote) [reply conans/model/conan_file.py]: Sounds good, but why is it related to this PR? shouldn't be another one with that stuff?

File: conan/tools/env/envtoolchain.py (hunk 2 above)
Comment (lasote) [reply conan/tools/env/envtoolchain.py]: The answer of "Why prepending like that" is  "because we cannot do better with the current model", right?

File: conans/model/conan_file.py (hunk 3 above)
Comment (memsharded) [reply conans/model/conan_file.py]: The ``EnvToolchain`` needs to explore the dependencies of the conanfile. It is implementing now access via:

```python
for dep in self._conanfile.dependencies.all:
```

Either we do an aggregation of all the ``new_env`` in a ``self._conanfile.deps_new_env_info``, or we allow the conanfile to visit its dependencies and aggregate the ``new_env`` directly itself. I think this second model is more structured, clean and scalable, and this is exactly the proposal of this PR: How the environment info can be propagated from recipes to their consumers.
